encrypted live migration

amphi opened the pull request:
This MR adds the possibility to encrypt the live migration with TLS.

I added new TLS variants to the `SocketStream` and `ReceiveListener`, so it should be more or less transparent. To enable the TLS encryption, the necessary parameters must be provided via `send-migration` and `receive-migration`.

I took inspiration from Qemu (https://www.qemu.org/docs/master/system/tls.html). `send-migration` and `receive-migration` both expect a path to a directory that contains the necessary certificates and keys. The names of the certificates and keys are hard coded. The TLS client (the side that does a `send-migration`) expects a `ca-cert.pem` file in the directory; the TLS server expects a `server-cert.pem` and a `server-key.pem`.

I tested everything using our libvirt-tests and it worked (although I did not verify that the data was really encrypted, so far I trust rustls).

## Todo

- [ ] update documentation

amphi committed:
vm-migration: add the client side of the TLS connection

TLS connections have a TLS server (the endpoint that listens for a
connection) and a TLS client (the endpoint that initiates the
connection). This commit adds the code for the client side, which will
be the source host.

On-behalf-of: SAP sebastian.eydam@sap.com
Signed-off-by: Sebastian Eydam <sebastian.eydam@cyberus-technology.de>

amphi committed:
vm-migration: add the server side of the TLS encryption

This is the TLS server side, which will be the live migration target.

On-behalf-of: SAP sebastian.eydam@sap.com
Signed-off-by: Sebastian Eydam <sebastian.eydam@cyberus-technology.de>

amphi committed:
vmm: remove AsRawFd trait for SocketStream, as it seems unnecessary

Also it seems like AsRawFd should be avoided
https://rust-lang.github.io/rfcs/3128-io-safety.html

On-behalf-of: SAP sebastian.eydam@sap.com
Signed-off-by: Sebastian Eydam <sebastian.eydam@cyberus-technology.de>

amphi committed:
vmm: add TLS variants to SocketStream and ReceiveListener

This allows (more or less) transparent usage of TLS encrypted TCP
connections.

On-behalf-of: SAP sebastian.eydam@sap.com
Signed-off-by: Sebastian Eydam <sebastian.eydam@cyberus-technology.de>

amphi committed:
vmm: use TLS encrypted live migration when TLS parameters are provided

For TLS we need certificates (and a key for the TLS server). This
commits adds parameters for that and encrypts the connection with TLS if
the necessary parameters are provided.

On-behalf-of: SAP sebastian.eydam@sap.com
Signed-off-by: Sebastian Eydam <sebastian.eydam@cyberus-technology.de>

phip1611 commented on vm-migration/src/lib.rs:
```diff
@@ -51,6 +52,9 @@ pub enum MigratableError {
     UnlockError(#[source] anyhow::Error),
+
+    #[error("TLS error")]
+    Tls(#[from] anyhow::Error),
```
please, if possible with managable effort: Please use a specific error type or an error enum. wrapping anyhow::Error like this is an antipattern and will also be [no longer accepted upstream](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/7432)

I know it is done everywhere in `enum MigrateableError` but it's definitely not good

amphi replied:
I changed that, please have another look. I am not sure whether I improved the situation :sweat_smile:

phip1611 replied:
I think it's good

phip1611 commented on vm-migration/src/tls.rs:
```diff
@@ -0,0 +1,107 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+use {
```
I'm a little surprised - is this what `cargo +nightly fmt --all` produced?

Don't get me wrong, I love this format! But I think this is not aligned with the .rustfmt.toml of this repo

amphi replied:
No, I wrote it like that after reading your comment in our Teams.

I (my editor) always just used `rustfmt`, which didn't change it. But now I see that `cargo +nightly fmt --all` reformats it.

amphi committed:
vm-migration: speed up ReadVolatile and WriteVolatile

The ReadVolatile and WriteVolatile implementations of TlsStream were
very slow, mainly because they allocated a large buffer on each
invocation. The TlsStreamWrapper carries a buffer that it uses for
ReadVolatile and WriteVolatile and that is allocated once on creation.

On-behalf-of: SAP sebastian.eydam@sap.com
Signed-off-by: Sebastian Eydam <sebastian.eydam@cyberus-technology.de>

phip1611 commented on Cargo.toml:
```diff
@@ -53,6 +53,7 @@ hypervisor = { path = "hypervisor" }
 libc = { workspace = true }
 log = { workspace = true, features = ["std"] }
 option_parser = { path = "option_parser" }
+rustls = "0.23.34"
```
please upgrade this to a workspace dependency and consume it then with `rustls = { workspace =  true; }`

amphi replied:
Fixed.

phip1611 commented on vm-migration/Cargo.toml:
```diff
@@ -7,6 +7,7 @@ version = "0.1.0"
 [dependencies]
 anyhow = { workspace = true }
 itertools = { workspace = true }
+rustls = "0.23.34"
```
same as above; workspace dependency

amphi replied:
Fixed.

amphi replied:
fixed.

phip1611 commented on vm-migration/src/tls.rs:
```diff
@@ -96,6 +103,52 @@ impl WriteVolatile for TlsStream {
+        let key_pem: PathBuf = [cert_dir, "server-key.pem"].iter().collect();
+        let key = PrivateKeyDer::from_pem_file(key_pem).unwrap();
+        let config = ServerConfig::builder()
+            .with_no_client_auth()
```
`with_no_client_auth()` sounds a little suspicious - is this how this library is supposed to be used?

amphi replied:
Otherwise we would require mTLS. Do we have an opinion on that?

phip1611 replied:
I don't know - sounds like a good thing to have. Maybe ask thomas and stefan

hertrste replied:
Multiple thoughts on that:
* QEMU supports mTLS for live migration via some `verify-peer=yes` parameter
* If we want to support it, it needs to be configurable
* we should reach out to the customer to see if they use mTLS for their current QEMU setup, and if they intend to use it
  - if possible, I would do mTLS as a follow up. This MR clearly brings a lot of benefit already and can already be upstreamed in this state also.

amphi replied:
So lets do the mTLS as a follow up then? Because it would also need integration work in libvirt, which means that all this takes even longer.

hertrste replied:
I would vote to do so yes. You can create a follow up ticket and reach out via Slack to see if mTLS should be supported and what the priority should be. I am very sure that TLS will be sufficient for GA.

amphi replied:
I opened https://github.com/cobaltcore-dev/cobaltcore/issues/367

phip1611 commented on vmm/src/api/mod.rs:
```diff
@@ -265,6 +265,9 @@ pub struct VmReceiveMigrationData {
     pub net_fds: Option<Vec<RestoredNetConfig>>,
+    /// Directory containing the TLS server certificate (server-cert.pem) and TLS server key (server-key.pem).
+    #[serde(default)]
+    pub tls_dir: Option<String>,
```
Please use PathBuf here which is the idiomatic way in Rust to handle owned paths

amphi replied:
Fixed.

phip1611 commented on src/bin/ch-remote.rs:
```diff
@@ -949,13 +958,15 @@ fn send_migration_data(
     downtime: u64,
     migration_timeout: u64,
     connections: NonZeroU32,
+    tls_dir: Option<&String>,
```
`&String` -> `&Path`

amphi replied:
Fixed.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -1233,15 +1234,14 @@ impl SendAdditionalConnections {
 
-        for n in 0..(connections.get() - 1) {
-            let socket = (match send_migration_socket(destination) {
+        for n in 0..(send_data_migration.connections.get() - 1) {
```
maybe a new binding: `let additiona_connections = send_data_migration.connections.get() - 1`?

amphi replied:
Why?

phip1611 replied:
To improve readability :) split one line into two

amphi replied:
Fixed.

phip1611 commented on vmm/src/lib.rs:
```diff
@@ -1385,17 +1385,31 @@ impl Drop for SendAdditionalConnections {
+            Ok(SocketStream::Tcp(socket))
+        } else {
+            info!("Live Migration will be encrypted using TLS.");
+            // The address may still contain a port. I think we should build something more robust to also handle IPv6.
```
there us `core::net::IpAddress` - we then can handle ipv4 and ipv6 - but are missing (ascii) host names

amphi replied:
But the parsing fails if it is an ip address that also contains a port (i tried that quickly with the compiler explorer).

phip1611 commented on vm-migration/src/tls.rs:
```diff
@@ -45,6 +45,22 @@ pub enum TlsStream {
 
+pub struct TlsStreamWrapper {
+    stream: TlsStream,
+    buf: Vec<u8>,
```
this buffer is to used to buffer read bytes from the underlying tls stream, yes?

amphi replied:
Exactly.

amphi replied:
I added a comment.

phip1611 commented on vm-migration/src/tls.rs:
```diff
@@ -45,6 +45,22 @@ pub enum TlsStream {
     Server(StreamOwned<ServerConnection, TcpStream>),
 }
 
+pub struct TlsStreamWrapper {
```
> The ReadVolatile and WriteVolatile implementations of TlsStream were
very slow

Is this something that you've debugged?

Please add this as comment to the `TlsStreamWrapper` type explaining why this wrapper is useful.

amphi replied:
Fixed.

phip1611 reviewed (changes requested):
You are on the right track! Left a few ramarks

phip1611 commented on src/bin/ch-remote.rs:
```diff
@@ -930,14 +939,15 @@ fn coredump_config(destination_url: &str) -> String {
 }
 
-fn receive_migration_data(url: &str) -> String {
+fn receive_migration_data(url: &str, tls_dir: Option<&PathBuf>) -> String {
```
```suggestion
fn receive_migration_data(url: &str, tls_dir: Option<&Path>) -> String {
```

`&Path` to `PathBuf` behaves similar like `&str` to `String`. You barely ever use `&String` and so you also barely ever use `&PathBuf`

**But I messed things up.** We take ownership anyway. sorry. So we can consume the owned type.

```suggestion
fn receive_migration_data(url: &str, tls_dir: Option<PathBuf>) -> String {
```

amphi replied:
Fixed.

phip1611 commented on src/bin/ch-remote.rs:
```diff
@@ -949,13 +959,15 @@ fn send_migration_data(
     downtime: u64,
     migration_timeout: u64,
     connections: NonZeroU32,
+    tls_dir: Option<&PathBuf>,
```
Ahhh, since we take ownership of the data here anyway, we could also use this. Sorry.

```suggestion
    tls_dir: Option<PathBuf>
```

amphi replied:
Fixed.

phip1611 commented on src/bin/ch-remote.rs:
```suggestion
    url: String,
```

as you need ownership anyway

amphi replied:
Fixed.

phip1611 commented on vm-migration/src/tls.rs:
```diff
@@ -0,0 +1,257 @@
+// with this wrapper.
+pub struct TlsStreamWrapper {
+    stream: TlsStream,
+    buf: Vec<u8>, // Used only in ReadVolatile and WriteVolatile
```
nit rust style
```suggestion
    // Used only in ReadVolatile and WriteVolatile
    buf: Vec<u8>, 
```

amphi replied:
Fixed.

phip1611 commented on vm-migration/src/tls.rs:
```diff
@@ -0,0 +1,257 @@
+    }
+}
+
+// Let's pray that nobody tries to read from or write to these fds.
```
😆  Could you please add some more higher-level context here?

amphi replied:
Did that.

phip1611 reviewed (changes requested):
Very nice! Left a few more nits

scholzp commented on vm-migration/src/tls.rs:
```diff
@@ -0,0 +1,133 @@
+#[derive(Error, Debug)]
+pub enum TlsError {
+    #[error(
+        "The provided input could not be parsed because it is not a syntactically-valid DNS Name."
```
Please note that errors, [by convention](https://doc.rust-lang.org/std/error/trait.Error.html), are usually written in lowercase and without trailing punctuation.

phip1611 replied:
That is true but in Cloud Hypervisor, by convention, we use normal sentence style.

amphi replied:
I left it as it is.

scholzp commented on vm-migration/src/tls.rs:
```diff
@@ -0,0 +1,133 @@
+// This TlsStream will be later encapsulated in a SocketStream. Thus it has to
+// implement the same traits. It is important that we never directly read from
+// or write to the TcpStream encapsulated in StreamOwned.
+pub enum TlsStream {
```
Please derive the `Debug` trait whenever possible

amphi replied:
Fixed.

scholzp reviewed (commented):
LGTM! Only some minor nits.

phip1611 reviewed (approved):
Super!

olivereanderson commented on vm-migration/src/tls.rs:
```diff
@@ -0,0 +1,261 @@
+    Server(StreamOwned<ServerConnection, TcpStream>),
+}
+
+// The TLS-Stream objects cannot read or write volatile, thus we need a buffer
+// between the VolatileSlice and the TLS stream (see ReadVolatile and
+// WriteVolatile implementations below). Allocating this buffer in these
+// function calls would make it very slow, thus we tie the buffer to the stream
+// with this wrapper.
+pub struct TlsStreamWrapper {
+    stream: TlsStream,
+    // Used only in ReadVolatile and WriteVolatile
+    buf: Vec<u8>,
+}
```
I am pretty sure that you can just wrap this in a [`BufReader`](https://doc.rust-lang.org/std/io/struct.BufReader.html) from the standard library. In fact, you may want to use that in more places as calling `Read::read` on the underlying TCP stream will always issue an expensive syscall.

phip1611 replied:
I think this is a great idea actually and might work. May I suggest the following. We merge it as is for the sake of progress and basti will look in a follow-up if it is feasible to use buf reader.

olivereanderson replied:
Yes, that is fine. Let's indeed try to merge often and iterate :)

amphi replied:
I don't think `BufReader` helps, because it does not implement `ReadVolatile` or `WriteVolatile`. So i'd still need this buffer.

olivereanderson replied:
You could implement `ReadVolatile` in a simpler way via the `BufReader`, but `WriteVolatile` would be harder indeed (as `BufReader` does not implement `Write`). I don't fully understand if it is a hard limitation that we need an instance that has both `Read` and `Write` or if we could split up our code somewhere, somehow.

Regardless, let's just merge this as-is and if we notice any performance issues, or technical debt down the line we can then explore this idea further.

phip1611 replied:
Anyhow, thanks for coming up with the idea of BufReader, oliver! It is always great to refresh the knowledge about cool helpers from the standard library

olivereanderson replied:
For the record: There is also [`BufWriter`](https://doc.rust-lang.org/std/io/struct.BufWriter.html), but unfortunately no `BufStream` which is what you would want here. The `tokio` crate does however [have that](https://docs.rs/tokio/latest/tokio/io/struct.BufStream.html).
